fix(upgrade): populate upgraded component's view before creating the controller

gkalpak opened the pull request:
**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] ~~Docs have been added / updated (for bug fixes / features)~~


**What kind of change does this PR introduce?** (check one with "x")
```
[x] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

**What is the current behavior?** (You can also link to an open issue here)
The relative order of the AngularJS compiling/linking operations is not similar to AngularJS's, resulting in inconsistent behavior for upgraded components (which makes upgrading to Angular less straight forward).


**What is the new behavior?**
The compiling/linking process of AngularJS is followed more closely.
Main differences:

- The components view is already populated when the controller is instantiated (and subsequent hooks are called).
- The correct DOM content is available when running the `$onChanges`, `$onInit`, `$doCheck` hooks.
   Previously, the "content children" were still present, not the "view children".
- The same for pre-linking.
- The template is compiled in the correct DOM context (e.g. has access to ancestors).
   Previously, it was compiled in isolation, inside a dummy element.


**Does this PR introduce a breaking change?** (check one with "x")
```
[ ] Yes
[x] No
```


**Other information**:
Fixes #13912.

For reference, here is the order of operations:

**Before**

1. Compile template
2. Instantiate controller
3. Hook: $onChanges
4. Hook: $onInit
5. Hook: $doCheck
6. Pre-linking
7. Collect content children
8. Insert compiled template
9. Linking
10. Post-linking
11. Hook: $postLink

**After**

1. Collect content children
2. Insert template
3. Compile template
4. Instantiate controller
5. Hook: $onChanges
6. Hook: $onInit
7. Hook: $doCheck
8. Pre-linking
9. Linking
10. Post-linking
11. Hook: $postLink

gkalpak committed:
fix(upgrade): populate upgraded component's view before creating the controller

Previously, the relative order of the AngularJS compiling/linking operations was
not similar to AngularJS's, resulting in inconsistent behavior for upgraded
components (which made upgrading to Angular less straight forward).

This commit fixes it, by following the compiling/linking process of AngularJS
more closely.

Main differences:

- The components view is already populated when the controller is instantiated
  (and subsequent hooks are called).
- The correct DOM content is available when running the `$onChanges`, `$onInit`,
  `$doCheck` hooks. Previously, the "content children" were still present, not
  the "view children".
- The same for pre-linking.
- The template is compiled in the correct DOM context (e.g. has access to
  ancestors). Previously, it was compiled in isolation, inside a dummy element.

For reference, here is the order of operations:

**Before**

1. Compile template
2. Instantiate controller
3. Hook: $onChanges
4. Hook: $onInit
5. Hook: $doCheck
6. Pre-linking
7. Collect content children
8. Insert compiled template
9. Linking
10. Post-linking
11. Hook: $postLink

**After**

1. Collect content children
2. Insert template
3. Compile template
4. Instantiate controller
5. Hook: $onChanges
6. Hook: $onInit
7. Hook: $doCheck
8. Pre-linking
9. Linking
10. Post-linking
11. Hook: $postLink

Fixes #13912

petebacondarwin commented on modules/@angular/upgrade/test/static/integration/upgrade_component_spec.ts:
```diff
@@ -1020,6 +1020,54 @@ export function main() {
 
+    describe('compiling', () => {
+      it('should compile the ng1 template in the correct DOM context', async(() => {
+           let grantParentNodeName: string;
```
should be `grandParentNodeName`?

petebacondarwin commented on modules/@angular/upgrade/test/static/integration/upgrade_component_spec.ts:
```diff
@@ -1020,6 +1020,54 @@ export function main() {
+             imports: [BrowserModule, UpgradeModule],
+             declarations: [Ng1ComponentAFacade, Ng2ComponentX],
+             entryComponents: [Ng2ComponentX],
+             schemas: [NO_ERRORS_SCHEMA],
```
We don't need `NO_ERRORS_SCHEMA`

petebacondarwin commented on modules/@angular/upgrade/test/static/integration/upgrade_component_spec.ts:
```diff
@@ -1253,6 +1301,55 @@ export function main() {
+
+           // Define `ng1Component`
+           const ng1Component: angular.IComponent = {
+             template: 'Hello, world!',
```
Given that you refer to "compiled" content here, is it worth making the template contain something dynamic, like an interpolation, rather than just a static string?

petebacondarwin commented on modules/@angular/upgrade/test/static/integration/upgrade_component_spec.ts:
```diff
@@ -1784,13 +1881,8 @@ export function main() {
-               $onChanges(changes: SimpleChanges) { controllerOnChangesB(changes); }
-             }
+             controller:
+                 class {$onChanges(changes: SimpleChanges) { controllerOnChangesB(changes); }}
```
Is this what our code formatter did to this line? Ugly!

gkalpak replied:
Apparently, it is both a formater and an obfuscator :persevere:

petebacondarwin commented on modules/@angular/upgrade/test/static/integration/upgrade_component_spec.ts:
```diff
@@ -1827,7 +1919,10 @@ export function main() {
+                                 .directive('ng2', downgradeComponent({component: Ng2Component}))
+                                 .run(($rootScope: angular.IRootScopeService) => {
+                                   Object.getPrototypeOf($rootScope)['$onChanges'] = scopeOnChanges;
+                                 });
```
What's going on here. Why this change?

gkalpak replied:
We need this to happen before `ngOnChanges` is run. Previously (as a convenience) I was doing this inside the controller, but now that the controller is instantiated after `ngOnChanges`, it is to late to patch `Scope.prototype`. So I moved it into a `run` block.

petebacondarwin commented on modules/@angular/upgrade/src/static/upgrade_component.ts:
```diff
@@ -413,19 +435,28 @@ export class UpgradeComponent implements OnInit, OnChanges, DoCheck, OnDestroy {
 
       // QUESTION: Do we want the ng1 component to call the function with `<value>` or with
```
I think we can remove this question now. It looks like we have agreed upon the former.

petebacondarwin commented on modules/@angular/upgrade/src/static/upgrade_component.ts:
```diff
@@ -321,6 +331,18 @@ export class UpgradeComponent implements OnInit, OnChanges, DoCheck, OnDestroy {
     return bindings;
   }
 
+  private getContentChildNodes(element: Element): Node[] {
```
Since this function has a side effect (of removing nodes from the `element`) then `get...` is not a good name, since it makes it sound like a pure getter function. Perhaps `extractContentChildNodes` or just `extractChildNodes` would be more appropriate?

petebacondarwin reviewed (changes requested):
Not quite finished but here is my review so far.

petebacondarwin commented on modules/@angular/upgrade/src/static/upgrade_component.ts:
```diff
@@ -137,24 +137,33 @@ export class UpgradeComponent implements OnInit, OnChanges, DoCheck, OnDestroy {
     this.$componentScope = $parentScope.$new(!!this.directive.scope);
 
+    this.tempBindingDestination = Object.create(null);
+    this.setupOutputs(this.tempBindingDestination);
```
Given that `setupOutputs` is only called here, why not change it to simply return this object and call it something like `createOutputs(): IBindingDestination`. So these two lines would become something like:

```ts
this.tempOutputHolder = this.setupOutputs();
```

petebacondarwin replied:
This would IMO make it easier to see what is going on here - i.e. that we need to be able to handle changes before the outputs get assigned to the binding destination (right?).

petebacondarwin commented on modules/@angular/upgrade/src/static/upgrade_component.ts:
```diff
@@ -137,24 +137,33 @@ export class UpgradeComponent implements OnInit, OnChanges, DoCheck, OnDestroy {
     this.bindingDestination = bindToController ? this.controllerInstance : this.$componentScope;
+    Object.keys(this.tempBindingDestination)
+        .forEach(key => this.bindingDestination[key] = this.tempBindingDestination[key]);
+    this.tempBindingDestination = null;
```
Then this bit of code would become something like:

```ts
    Object.keys(this.tempOutputHolder)
        .forEach(key => this.bindingDestination[key] = this.tempOutputHolder[key]);
    this.tempOutputHolder = null;
```

(It is a shame we probably cannot use `Object.assign` here.)

petebacondarwin commented on modules/@angular/upgrade/src/static/upgrade_component.ts:
```diff
@@ -95,10 +95,11 @@ export class UpgradeComponent implements OnInit, OnChanges, DoCheck, OnDestroy {
+  private controllerInstance: IControllerInstance;
+  private bindingDestination: IBindingDestination;
+  private tempBindingDestination: IBindingDestination;
+  private pendingChanges: SimpleChanges;
```
I think some comments about these two new properties to explain their purpose to future readers would be helpful here.

petebacondarwin reviewed (changes requested):
The behaviour seems correct. I have a few minor suggestions about the implementation, naming and commenting.

gkalpak commented:
@petebacondarwin: PTAL
I addressed your comments and made another change redarding `setupOutputs`.
I realized we were doing two thing in `setupOutpus`:

1. Define `EventEmitters` for the outputs on the ng2 facade compoenent. (This needed to happen in the component constructor, or Angular would complain).
2. Set up the forwarding of the `&` bindings from the ng1 component to the ng2 component facade's `EventEmitter`. (This is not needed in the constructor, but it requires the binding destination.)

So, I kept the first part in the constructor and moved the second part to `ngOnInit` (after the ng1 controller has been created).

gkalpak commented:
(The Travis failure is unrelated btw. Inherited it from master.)

petebacondarwin commented on modules/@angular/upgrade/src/static/upgrade_component.ts:
```diff
@@ -95,10 +95,13 @@ export class UpgradeComponent implements OnInit, OnChanges, DoCheck, OnDestroy {
+  private bindingDestination: IBindingDestination;
+
+  // We will be instantiating the controller in the `ngOnInit` hook, when the first `ngOnChanges`
+  // will have been already triggered. We store the `SimpleChanges` and to "play them back" later.
```
> We store the `SimpleChanges` and ~~to~~ "play them back" later.

Remove the "to"

petebacondarwin commented on modules/@angular/upgrade/src/static/upgrade_component.ts:
```diff
@@ -405,32 +426,42 @@ export class UpgradeComponent implements OnInit, OnChanges, DoCheck, OnDestroy {
   }
 
   private setupOutputs() {
```
could we call this `createOutputs` or something similar?

gkalpak replied:
How about `initializeOutputs()` (as it basically initializes the output properties to `new EventEmitter()`)?

petebacondarwin commented on modules/@angular/upgrade/src/static/upgrade_component.ts:
```diff
@@ -405,32 +426,42 @@ export class UpgradeComponent implements OnInit, OnChanges, DoCheck, OnDestroy {
+        });
+  }
 
+  private setupExpressionBindings() {
```
And then call this `bindOutputs` or something similar?

petebacondarwin reviewed (approved):
I made some naming suggestions but they are not blockers on landing this.

gkalpak commented:
I addressed the comments and squashed.

petebacondarwin commented:
LGTM
